core/internal/cltest/mocks.go

```diff
@@ -334,21 +334,30 @@ func fwdHeaders(actual, mock interface{}) {
 
 // MockSubscription a mock subscription
 type MockSubscription struct {
-	name    string
-	channel interface{}
-	Errors  chan error
+	mut          sync.Mutex
+	name         string
+	channel      interface{}
+	unsubscribed bool
+	Errors       chan error
 }
 
 // EmptyMockSubscription return empty MockSubscription
-func EmptyMockSubscription() MockSubscription {
-	return MockSubscription{Errors: make(chan error, 1), channel: make(chan struct{})}
+func EmptyMockSubscription() *MockSubscription {
+	return &MockSubscription{Errors: make(chan error, 1), channel: make(chan struct{})}
 }
 
 // Err returns error channel from mes
-func (mes MockSubscription) Err() <-chan error { return mes.Errors }
+func (mes *MockSubscription) Err() <-chan error { return mes.Errors }
 
 // Unsubscribe closes the subscription
-func (mes MockSubscription) Unsubscribe() {
+func (mes *MockSubscription) Unsubscribe() {
+	mes.mut.Lock()
```

Comment:
There's probably a way to do this with pure channels but I'll leave that to @spook.

The mutex looks safe to me.

Reply:
Mutex's are pretty common in Golang world, specially if the code isn't critically high performant.